# Introducción a DataFrames en Pandas
En esta sección aprenderás los conceptos básicos sobre DataFrames, la estructura fundamental para el análisis de datos en Python usando la librería Pandas.

In [162]:
import pandas as pd
import os

In [163]:
#ruta del archivo actual

curr_path = os.getcwd()
print(curr_path)

datasets_path = os.path.join(curr_path, "..", "All_Datasets")
print(datasets_path)

c:\Users\MATEOCRACK\Desktop\familia\santi\factultade\proc_limpieza_visualizazcion\teoria
c:\Users\MATEOCRACK\Desktop\familia\santi\factultade\proc_limpieza_visualizazcion\teoria\..\All_Datasets


## ¿Qué es un DataFrame?
Un **DataFrame** es una estructura de datos bidimensional (tabla) que permite almacenar y manipular información en filas y columnas, similar a una hoja de cálculo de Excel.
- Cada columna puede tener un tipo de dato diferente (números, texto, fechas, etc.).
- Los valores faltantes se representan con `NaN` (*Not a Number*), y es importante saber cómo gestionarlos.
- Tanto las filas como las columnas tienen etiquetas (índices) que facilitan el acceso y manipulación de los datos.
- Pandas ofrece métodos comunes para Series y DataFrames, como `info()`, pero también métodos exclusivos para cada tipo.

In [210]:

nba = pd.read_csv(datasets_path + "/nba.csv", sep=";")
display(nba)

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN


In [211]:
nba.tail()

,Name,Team,Position,Height,Weight,College,Salary
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN
590,Delon Wright,Washington Wizards,G,6-may,185.0,Utah,8195122.0


In [212]:
nba.info()

<class 'pandas.DataFrame'>
RangeIndex: 591 entries, 0 to 590
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      591 non-null    str    
 1   Team      591 non-null    str    
 2   Position  584 non-null    str    
 3   Height    585 non-null    str    
 4   Weight    584 non-null    float64
 5   College   578 non-null    str    
 6   Salary    488 non-null    float64
dtypes: float64(2), str(5)
memory usage: 32.4 KB


In [213]:
s = pd.Series(["1", 2,3,4,5])

print(s.dtypes)
print(s.index)
print(s.values)


object
RangeIndex(start=0, stop=5, step=1)
['1' 2 3 4 5]


Los dataframes estan compuestos por:
- **INDEX**:En este caso los indices del dataframe son numeros en el rango 0, 591, step 1. el objeto nba.index es una lista de la forma [0, 1, .., n] (en este caso). Se pueden acceder a estos elementos
$$$$
- **VALUES**: Los valores son array de datos, asociados a cada indices. el objeto nba.values es una lista de listas, pudiendo acceder por indices a cada valor.
$$$$

- **COLUMN**: Se agrega un tipo algo especial respecto a las series, las columnas, que en teoria deberian ser la misma cantidad de elementos que las observaciones
___________
$$$$

**Mención honorífica: AXIS**
Los axis son los ejes dimensionales que se manejan nuestros datos y que representan.
- Series: Teniamos una dimension de los datos y sus axis eran los indices
- Dataframe: tenemos dos dimensiones y los ejes son dos, indices y columnas.

nba.axis da una lista con dos elementos:
- nba.axis[0]: representa los indices (fila)
- nba.axis[1]: representa las columnas


In [214]:
print(nba.values)
print()
print(nba.index)
print()
print(nba.columns)

print()
print()

print(s.axes)
print(nba.axes)

[['Saddiq Bey' 'Atlanta Hawks' 'F' ... 215.0 'Villanova' 4556983.0]
 ['Bogdan Bogdanovic' 'Atlanta Hawks' 'G' ... 225.0 'Fenerbahce'
  18700000.0]
 ['Kobe Bufkin' 'Atlanta Hawks' 'G' ... 195.0 'Michigan' 4094244.0]
 ...
 ['Landry Shamet' 'Washington Wizards' 'G' ... 190.0 'Wichita State'
  10250000.0]
 ['Tristan Vukcevic' 'Washington Wizards' 'F' ... 220.0 'Real Madrid' nan]
 ['Delon Wright' 'Washington Wizards' 'G' ... 185.0 'Utah' 8195122.0]]

RangeIndex(start=0, stop=591, step=1)

Index(['Name', 'Team', 'Position', 'Height', 'Weight', 'College', 'Salary'], dtype='str')


[RangeIndex(start=0, stop=5, step=1)]
[RangeIndex(start=0, stop=591, step=1), Index(['Name', 'Team', 'Position', 'Height', 'Weight', 'College', 'Salary'], dtype='str')]


In [215]:
#dimension
print(s.shape)
print(nba.shape)

print()

#tipos de datos
print(s.dtypes)
print(nba.dtypes)

(5,)
(591, 7)

object
Name            str
Team            str
Position        str
Height          str
Weight      float64
College         str
Salary      float64
dtype: object


In [216]:
print(s.hasnans)
#print(nba.hasnans) No existe tal cosa(tampoco sabia que series tenia)

False


In [217]:
print(s.info())

print()

print(nba.info())

<class 'pandas.Series'>
RangeIndex: 5 entries, 0 to 4
Series name: None
Non-Null Count  Dtype 
--------------  ----- 
5 non-null      object
dtypes: object(1)
memory usage: 172.0+ bytes
None

<class 'pandas.DataFrame'>
RangeIndex: 591 entries, 0 to 590
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      591 non-null    str    
 1   Team      591 non-null    str    
 2   Position  584 non-null    str    
 3   Height    585 non-null    str    
 4   Weight    584 non-null    float64
 5   College   578 non-null    str    
 6   Salary    488 non-null    float64
dtypes: float64(2), str(5)
memory usage: 32.4 KB
None


## Diferencias entre métodos compartidos
Algunos métodos de Pandas funcionan tanto en Series como en DataFrames, pero su comportamiento puede variar:
- **`sum` en Series:** Suma todos los valores de la serie.
- **`sum` en DataFrame:** Por defecto suma los valores de cada columna (por filas).
Puedes usar el parámetro `axis` para cambiar el comportamiento:
- `axis=0` o "index" (comportamiento por defecto).
- `axis=1` o "columns"

Con el atributo `index_col` le podemos decir que columna usar de indice si viene. Si hay indices duplicados no falla, pero cambia la interaccion con el acceso por indices:
- Si el indice es unico, el acceso da una `Serie` de datos con indices las columnas del dataframe
- Si el indice no es unico, el resultado es un `Dataframe`

In [172]:
revenue = pd.read_csv(datasets_path+"/revenue.csv", index_col="Date",sep=";")
#revenue_index = pd.read_csv(datasets_path+"/revenue.csv",sep = ";")
display(revenue)

print(type(revenue.loc["1/10/2026"]))
print(revenue.loc["1/10/2026"].index)
print(type(revenue.loc["1/1/2026"]))
print(revenue.loc["1/1/2026"].index)


,New York,Los Angeles,Miami
Date,,,
1/1/2026,985,122,499
1/2/2026,738,788,534
1/3/2026,14,20,933
1/4/2026,730,904,885
1/5/2026,114,71,253
1/6/2026,936,502,497
1/7/2026,123,996,115
1/8/2026,935,492,886
1/9/2026,846,954,823


<class 'pandas.DataFrame'>
Index(['1/10/2026', '1/10/2026'], dtype='str', name='Date')
<class 'pandas.Series'>
Index(['New York', 'Los Angeles', 'Miami'], dtype='str')


Algunos metodos sobre series utilizando axis: Axis indica sobre las operacion que se va aplicar, lo que "desaparece" luego de aplicar la operacion
- Axis = "index" o axis=0, desaparecen los indices(renglones). La suma retorna una serie

In [173]:
sum = revenue.sum(axis = 0)
display(sum)
print(type(sum))#serieeeeeeeeee

print(revenue.sum(axis = "columns", )) #o axis =1


New York       5530
Los Angeles    5419
Miami          5857
dtype: int64

<class 'pandas.Series'>
Date
1/1/2026     1606
1/2/2026     2060
1/3/2026      967
1/4/2026     2519
1/5/2026      438
1/6/2026     1935
1/7/2026     1234
1/8/2026     2313
1/9/2026     2623
1/10/2026     555
1/10/2026     556
dtype: int64


## Selección de columnas en un DataFrame
- Para seleccionar una sola columna, puedes usar:
  - `df.columna` (solo si el nombre no tiene espacios).
  - `df["columna"]` (siempre funciona).
  Esto devuelve una **Series**.
- Para seleccionar varias columnas, usa una lista:
  - `df[["col1", "col2"]]`
  Esto devuelve un nuevo **DataFrame**.
- Si necesitas modificar los datos sin afectar el original, utiliza `.copy()` para crear una copia independiente.

_____________________

**Acceso a columnas**
Las columnas se pueden acceder con 
- Notacion punto `df.column`: Retorna serie con los datos
- Notacion corchetes `df[[column]]`: Podemos identificar mas de una columna. Si es una, es serie, si son varias es dataframe.


$$$$

**Acceso por indices:** Tenemos que acceder sabiendo los indices que tiene nuestro dataframe:
- df.loc[indice, columna]: usado como indice columna:
- df.iloc[posx, posy]: usado como pocision en el dataframe


In [174]:
#Revisar si tenemos una vista o una copia
nba = pd.read_csv(datasets_path + "/nba.csv", sep = ";")
nba.head()

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0


In [175]:
# ================ acceso por columna================
print("--------------------------------COLUMNAS---------------------")
#acceso notacion punto
display(nba.Team)
print(type(nba.Team))

#acceso corchetes.
display(nba["Team"])
print(type(nba["Team"]))


col = ["Team","Salary","Name"]
display(nba[col])
print(type(nba[col]))



# ============== Acceso por fila ====================
print("--------------------------------INDICES---------------------")
display(nba.iloc[0])
display(nba.loc[0])



--------------------------------COLUMNAS---------------------


0           Atlanta Hawks
1           Atlanta Hawks
2           Atlanta Hawks
3           Atlanta Hawks
4           Atlanta Hawks
              ...        
586    Washington Wizards
587    Washington Wizards
588    Washington Wizards
589    Washington Wizards
590    Washington Wizards
Name: Team, Length: 591, dtype: str

<class 'pandas.Series'>


0           Atlanta Hawks
1           Atlanta Hawks
2           Atlanta Hawks
3           Atlanta Hawks
4           Atlanta Hawks
              ...        
586    Washington Wizards
587    Washington Wizards
588    Washington Wizards
589    Washington Wizards
590    Washington Wizards
Name: Team, Length: 591, dtype: str

<class 'pandas.Series'>


,Team,Salary,Name
0,Atlanta Hawks,4556983.0,Saddiq Bey
1,Atlanta Hawks,18700000.0,Bogdan Bogdanovic
2,Atlanta Hawks,4094244.0,Kobe Bufkin
3,Atlanta Hawks,20616000.0,Clint Capela
4,Atlanta Hawks,2581522.0,Bruno Fernando
...,...,...,...
586,Washington Wizards,27955357.0,Jordan Poole
587,Washington Wizards,1719864.0,Ryan Rollins
588,Washington Wizards,10250000.0,Landry Shamet
589,Washington Wizards,NaN,Tristan Vukcevic


<class 'pandas.DataFrame'>
--------------------------------INDICES---------------------


Name           Saddiq Bey
Team        Atlanta Hawks
Position                F
Height              6-jul
Weight              215.0
College         Villanova
Salary          4556983.0
Name: 0, dtype: object

Name           Saddiq Bey
Team        Atlanta Hawks
Position                F
Height              6-jul
Weight              215.0
College         Villanova
Salary          4556983.0
Name: 0, dtype: object

In [176]:
nba

#accedo a indice cero
display(nba.loc[0])
print()

#ene ste caso indices y posiciones coinciden pero bueno
display(nba.iloc[0])
print()

#acceder x,y

print(nba.loc[0, "Team"]) #es el type del dato

print(nba.iloc[0, 1]) #mas incomodo


Name           Saddiq Bey
Team        Atlanta Hawks
Position                F
Height              6-jul
Weight              215.0
College         Villanova
Salary          4556983.0
Name: 0, dtype: object

Name           Saddiq Bey
Team        Atlanta Hawks
Position                F
Height              6-jul
Weight              215.0
College         Villanova
Salary          4556983.0
Name: 0, dtype: object


Atlanta Hawks
Atlanta Hawks


Python trabaja siempre por referencia, al querer guardarme algo en particular, hay que hacer una copia siempre para no modificar la referencia.

In [177]:
names = nba["Name"].copy()
names.iloc[2] = "mbappe"

names #no cambio el dataset original

0             Saddiq Bey
1      Bogdan Bogdanovic
2                 mbappe
3           Clint Capela
4         Bruno Fernando
             ...        
586         Jordan Poole
587         Ryan Rollins
588        Landry Shamet
589     Tristan Vukcevic
590         Delon Wright
Name: Name, Length: 591, dtype: str

## Cómo agregar una nueva columna a un DataFrame
- Puedes crear una nueva columna asignando una Serie o un valor:
  - `df["nueva_columna"] = valores`
- El método `insert` te permite ubicar la columna en una posición específica:
  - `df.insert(loc=índice, column="nombre", value=valores)`
- Puedes realizar operaciones entre columnas para crear nuevas variables (broadcasting).

________
**Agregar columnas:**
- mediante indices: Si vos escribis sobre `df["algo inexistente"] = values` entonces se crea la columna automaticamente y se agrega al final
- mediante su posicion x,y: Indicandole la posicion columna que quiere desplazar para poner esas columnas.

In [178]:
nba = pd.read_csv(datasets_path + "/nba.csv", sep = ";")
nba

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN


In [179]:
nba["doubled_SalarY"] = nba["Salary"] * 2
display(nba)

,Name,Team,Position,Height,Weight,College,Salary,doubled_SalarY
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0,9113966.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0,37400000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0,8188488.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0,41232000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0,5163044.0
...,...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0,55910714.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0,3439728.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0,20500000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN,NaN


In [180]:

#concatenando valores
nba["Nombre y Equipo"] = nba["Name"] +" - "+ nba["Team"]
display(nba)


,Name,Team,Position,Height,Weight,College,Salary,doubled_SalarY,Nombre y Equipo
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0,9113966.0,Saddiq Bey - Atlanta Hawks
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0,37400000.0,Bogdan Bogdanovic - Atlanta Hawks
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0,8188488.0,Kobe Bufkin - Atlanta Hawks
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0,41232000.0,Clint Capela - Atlanta Hawks
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0,5163044.0,Bruno Fernando - Atlanta Hawks
...,...,...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0,55910714.0,Jordan Poole - Washington Wizards
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0,3439728.0,Ryan Rollins - Washington Wizards
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0,20500000.0,Landry Shamet - Washington Wizards
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN,NaN,Tristan Vukcevic - Washington Wizards


In [181]:
#sabiendo las posiciones absolutas donde queres insertarla
# 0 name, 1, team, 2 position
nba.insert(loc=3, column="League", value="National Basketball Association")
nba

,Name,Team,Position,League,Height,Weight,College,Salary,doubled_SalarY,Nombre y Equipo
0,Saddiq Bey,Atlanta Hawks,F,National Basketball Association,6-jul,215.0,Villanova,4556983.0,9113966.0,Saddiq Bey - Atlanta Hawks
1,Bogdan Bogdanovic,Atlanta Hawks,G,National Basketball Association,6-may,225.0,Fenerbahce,18700000.0,37400000.0,Bogdan Bogdanovic - Atlanta Hawks
2,Kobe Bufkin,Atlanta Hawks,G,National Basketball Association,6-may,195.0,Michigan,4094244.0,8188488.0,Kobe Bufkin - Atlanta Hawks
3,Clint Capela,Atlanta Hawks,C,National Basketball Association,6-oct,256.0,Elan Chalon,20616000.0,41232000.0,Clint Capela - Atlanta Hawks
4,Bruno Fernando,Atlanta Hawks,F-C,National Basketball Association,6-oct,240.0,Maryland,2581522.0,5163044.0,Bruno Fernando - Atlanta Hawks
...,...,...,...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,National Basketball Association,6-abr,194.0,Michigan,27955357.0,55910714.0,Jordan Poole - Washington Wizards
587,Ryan Rollins,Washington Wizards,G,National Basketball Association,6-mar,180.0,Toledo,1719864.0,3439728.0,Ryan Rollins - Washington Wizards
588,Landry Shamet,Washington Wizards,G,National Basketball Association,6-abr,190.0,Wichita State,10250000.0,20500000.0,Landry Shamet - Washington Wizards
589,Tristan Vukcevic,Washington Wizards,F,National Basketball Association,6-oct,220.0,Real Madrid,NaN,NaN,Tristan Vukcevic - Washington Wizards


## operaciones sobre los datos:
- Podemos `sum()`, `mult()`,  `sub()`, `div()` por una constante los datos o aplicar la operacion directamente (en caso de ser numericos)
- Agregar string en caso de ser texto. 

In [182]:
nba["doubled_SalarY"] = nba["Salary"] * 2
display(nba)

nba["Salary"] - 5000000
nba["Salary"].sub(5000000)

nba["New Salary"] = nba["Salary"].sub(5000000)

,Name,Team,Position,League,Height,Weight,College,Salary,doubled_SalarY,Nombre y Equipo
0,Saddiq Bey,Atlanta Hawks,F,National Basketball Association,6-jul,215.0,Villanova,4556983.0,9113966.0,Saddiq Bey - Atlanta Hawks
1,Bogdan Bogdanovic,Atlanta Hawks,G,National Basketball Association,6-may,225.0,Fenerbahce,18700000.0,37400000.0,Bogdan Bogdanovic - Atlanta Hawks
2,Kobe Bufkin,Atlanta Hawks,G,National Basketball Association,6-may,195.0,Michigan,4094244.0,8188488.0,Kobe Bufkin - Atlanta Hawks
3,Clint Capela,Atlanta Hawks,C,National Basketball Association,6-oct,256.0,Elan Chalon,20616000.0,41232000.0,Clint Capela - Atlanta Hawks
4,Bruno Fernando,Atlanta Hawks,F-C,National Basketball Association,6-oct,240.0,Maryland,2581522.0,5163044.0,Bruno Fernando - Atlanta Hawks
...,...,...,...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,National Basketball Association,6-abr,194.0,Michigan,27955357.0,55910714.0,Jordan Poole - Washington Wizards
587,Ryan Rollins,Washington Wizards,G,National Basketball Association,6-mar,180.0,Toledo,1719864.0,3439728.0,Ryan Rollins - Washington Wizards
588,Landry Shamet,Washington Wizards,G,National Basketball Association,6-abr,190.0,Wichita State,10250000.0,20500000.0,Landry Shamet - Washington Wizards
589,Tristan Vukcevic,Washington Wizards,F,National Basketball Association,6-oct,220.0,Real Madrid,NaN,NaN,Tristan Vukcevic - Washington Wizards


## Revisión del método `value_counts`
El método `value_counts` permite contar cuántas veces aparece cada valor único en una **Series**. Es útil para analizar la frecuencia de categorías o valores en tus datos.

In [183]:
nba = pd.read_csv(datasets_path + "/nba.csv", sep=";")
nba.head()

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0


In [184]:
serie = nba["Position"].value_counts() #da una serie con los valores unicos y su cantidad repeuta
display(serie)


normalized_serie = nba["Position"].value_counts(normalize=True)
normalized_serie = normalized_serie * 100

display(normalized_serie)




Position
G      229
F      187
C       47
G-F     46
F-C     37
C-F     23
F-G     15
Name: count, dtype: int64

Position
G      39.212329
F      32.020548
C       8.047945
G-F     7.876712
F-C     6.335616
C-F     3.938356
F-G     2.568493
Name: proportion, dtype: float64

In [185]:
#puedo redondear los datos
normalized_serie = normalized_serie.round(2)
display(normalized_serie)

Position
G      39.21
F      32.02
C       8.05
G-F     7.88
F-C     6.34
C-F     3.94
F-G     2.57
Name: proportion, dtype: float64

## Manejo de valores faltantes: eliminar filas
Para limpiar tu DataFrame de valores faltantes, puedes usar el método `dropna`:
- Por defecto elimina filas con *cualquier* valor `NaN`.
- Usa el parámetro `how="all"` para eliminar filas donde *todos* los valores son `NaN`.
- Usa `subset` para especificar columnas a evaluar.
Esto ayuda a preparar los datos para análisis más precisos.

In [186]:
nba = pd.read_csv(datasets_path + "/nba.csv", sep = ";")
nba.head()

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0


In [187]:
nba.info()

print(nba.shape)

<class 'pandas.DataFrame'>
RangeIndex: 591 entries, 0 to 590
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      591 non-null    str    
 1   Team      591 non-null    str    
 2   Position  584 non-null    str    
 3   Height    585 non-null    str    
 4   Weight    584 non-null    float64
 5   College   578 non-null    str    
 6   Salary    488 non-null    float64
dtypes: float64(2), str(5)
memory usage: 32.4 KB
(591, 7)


In [188]:
#nba.dropna(how = "any") elimina si falta algun dato
#nba.dropna(how="all") elimina si faltan todos los datos

nba.dropna(subset = ["Position", "Height"], how = "all") #elimino todos los que le faltan ambos

nba.dropna(subset = ["Salary"], how = "any")

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
...,...,...,...,...,...,...,...
585,Eugene Omoruyi,Washington Wizards,F,6-jun,235.0,Oregon,559782.0
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0


## Rellenar valores faltantes con `fillna` (Imputacion de datos)
El método `fillna` reemplaza valores `NaN` con el argumento proporcionado.
- Puedes usarlo en **DataFrames** y **Series**.
- Devuelve una copia, no afecta el objeto original a menos que lo reasignes.
- Es útil para preparar los datos antes de análisis o visualización.


Observaciones: DEVUELVE COPIA, entonces es necesario indicarle el dataset para que aplique el cambio. Tiene inplace

In [189]:
nba = pd.read_csv(datasets_path+"/nba.csv", sep = ";").dropna(how="all")
nba
#Con all no borramos las rows que contienen data + NaN

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN


In [190]:
#Reemplazamos los valores con 0 con fillna
nba_filled = nba.fillna(0) # no modifica el dataframe original, da una copia
display(nba_filled)

#Nos enfocamos en una sola columna y rellenamos con 0 los NaN
nba["Salary"] = nba["Salary"].fillna(nba["Salary"].mean())
nba["Team"] = nba["Team"].fillna("Por Asignar")

display(nba[nba.Team=="Por Asignar"])
#En el caso de college queremos poner "desconocido" y no un 0.0
nba["College"] = nba["College"].fillna(value="Unknown")
display(nba[nba.College=="Unknown"])



,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,0.0


,Name,Team,Position,Height,Weight,College,Salary


,Name,Team,Position,Height,Weight,College,Salary
15,Miles Norris,Atlanta Hawks,NaN,NaN,NaN,Unknown,5.597820e+05
74,Onuralp Bitim,Chicago Bulls,F,6-jun,215.0,Unknown,5.597820e+05
129,Jordan Walker,Dallas Mavericks,NaN,NaN,NaN,Unknown,1.119563e+06
134,Souley Boum,Denver Nuggets,NaN,NaN,NaN,Unknown,1.119563e+06
138,Armaan Franklin,Denver Nuggets,NaN,NaN,NaN,Unknown,1.119563e+06
202,Joshua Obiesie,Houston Rockets,G-F,6-jun,195.0,Unknown,1.119563e+06
274,Tarik Biberovic,Memphis Grizzlies,F,6-jul,218.0,Unknown,9.218978e+06
299,Caleb Daniels,Miami Heat,NaN,NaN,NaN,Unknown,1.119563e+06
323,Jazian Gortman,Milwaukee Bucks,G,6-feb,184.0,Unknown,1.119563e+06
381,Jaylen Martin,New York Knicks,G,6-jun,216.0,Unknown,5.597820e+05


## Conversión de tipos de datos con `astype`
El método `astype` permite cambiar el tipo de datos de una columna, por ejemplo, de `float` a `int` o a `category`.
- Es importante asegurarse de que no haya valores faltantes antes de convertir.
- Usar el tipo `category` es útil para columnas con pocos valores únicos, optimizando el uso de memoria.
- El atributo `dtypes` muestra los tipos de las columnas en el DataFrame.


Observaciones: DEVUELVE COPIA, entonces es necesario indicarle el dataset para que aplique el cambio. No tiene inplace

In [191]:
nba = pd.read_csv(datasets_path + "/nba.csv", sep = ";").dropna(how = "all") #Toda la fila debe tener N/A
nba["Salary"] = nba["Salary"].fillna(0)
nba["Weight"] = nba["Weight"].fillna(0)
nba

,Name,Team,Position,Height,Weight,College,Salary
0,Saddiq Bey,Atlanta Hawks,F,6-jul,215.0,Villanova,4556983.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
...,...,...,...,...,...,...,...
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,0.0


In [192]:
display(nba.dtypes) #retorna una serie

print(nba.info())

Name            str
Team            str
Position        str
Height          str
Weight      float64
College         str
Salary      float64
dtype: object

<class 'pandas.DataFrame'>
RangeIndex: 591 entries, 0 to 590
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      591 non-null    str    
 1   Team      591 non-null    str    
 2   Position  584 non-null    str    
 3   Height    585 non-null    str    
 4   Weight    591 non-null    float64
 5   College   578 non-null    str    
 6   Salary    591 non-null    float64
dtypes: float64(2), str(5)
memory usage: 32.4 KB
None


In [193]:
copia = nba["Salary"].astype("int") #genera siempre una copia!
copia.dtype #funciona
copia2=nba["Salary"].astype(int) #tambien da copia
copia2.dtype

nba["Salary"] = nba["Salary"].astype("int")


## El tipo `category` en Pandas
- El tipo `category` es ideal para columnas con pocos valores únicos, como posiciones o equipos.
- Usar `category` reduce el uso de memoria y mejora el rendimiento.
- Puedes contar los valores únicos con `nunique` y convertir una columna con `astype("category")`.


Observaciones: Los tipos de datos de categoria son mas pesados en memoria!!!

In [194]:
nba = pd.read_csv(datasets_path + "/nba.csv", sep = ";").dropna(how = "all")
nba.tail()

,Name,Team,Position,Height,Weight,College,Salary
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN
590,Delon Wright,Washington Wizards,G,6-may,185.0,Utah,8195122.0


In [195]:
aa = nba["Team"].unique() #me da una lista de valores unicos
bb = nba.Team.nunique() #meda una lista de la cantidad de valroes unicos
print(bb)

nunique = nba.nunique() #Valores unicos por columna
display(nunique) # es una serie

30


Name        591
Team         30
Position      7
Height       20
Weight       93
College     182
Salary      298
dtype: int64

In [196]:
print(nba.info())
#Mirar la memory usage: 36.9+ KB

nba["Position"] = nba["Position"].astype("category")
nba["Team"] = nba["Team"].astype("category")

print()
print(nba.info())

<class 'pandas.DataFrame'>
RangeIndex: 591 entries, 0 to 590
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Name      591 non-null    str    
 1   Team      591 non-null    str    
 2   Position  584 non-null    str    
 3   Height    585 non-null    str    
 4   Weight    584 non-null    float64
 5   College   578 non-null    str    
 6   Salary    488 non-null    float64
dtypes: float64(2), str(5)
memory usage: 32.4 KB
None

<class 'pandas.DataFrame'>
RangeIndex: 591 entries, 0 to 590
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype   
---  ------    --------------  -----   
 0   Name      591 non-null    str     
 1   Team      591 non-null    category
 2   Position  584 non-null    category
 3   Height    585 non-null    str     
 4   Weight    584 non-null    float64 
 5   College   578 non-null    str     
 6   Salary    488 non-null    float64 
dtypes: category(2), float64(2), str(3)
memory u

## Ordenar un DataFrame con `sort_values`
Puedes ordenar tu DataFrame por una o varias columnas usando el método `sort_values`:
- Usa el parámetro `by` para especificar la columna o columnas.
- Personaliza el orden con `ascending` (True por defecto).
- Ajusta la posición de los `NaN` con `na_position`.

Observaciones. Le podemos pasar inplace para que no retorne una copia.

In [197]:
nba = pd.read_csv(datasets_path + "/nba.csv", sep = ";")

In [198]:
nba.sort_values(by = "Salary", ascending = False, na_position = "last", inplace = True)
# nba = nba.sort_values(by = "Salary", ascending = False, na_position = "last") Lo mismo que el inplace = True
display(nba)

nba = nba.sort_values(by=["Team","Salary"], ascending = True, na_position = "last") 
display(nba)

nba = nba.sort_values(by=["Team","Salary"], ascending = [True,False], na_position = "last")
display(nba)


# Como existe sort_values, tambien existe sort_index

nba = nba.sort_index(ascending=False)
nba


,Name,Team,Position,Height,Weight,College,Salary
175,Stephen Curry,Golden State Warriors,G,6-feb,185.0,Davidson,51915615.0
461,Kevin Durant,Phoenix Suns,F,6-oct,240.0,Texas,47649433.0
261,LeBron James,Los Angeles Lakers,F,6-sept,250.0,St. Vincent-St. Mary HS (OH),47607350.0
436,Joel Embiid,Philadelphia 76ers,C-F,7-0,280.0,Kansas,47607350.0
145,Nikola Jokic,Denver Nuggets,C,6-nov,284.0,Mega Basket,47607350.0
...,...,...,...,...,...,...,...
544,Dennis Schroder,Toronto Raptors,G,6-ene,172.0,Braunschweig,NaN
547,Gary Trent Jr.,Toronto Raptors,G-F,6-may,209.0,Duke,NaN
578,Taj Gibson,Washington Wizards,F,6-sept,232.0,Southern California,NaN
584,Kendrick Nunn,Washington Wizards,G,6-mar,190.0,Oakland,NaN


,Name,Team,Position,Height,Weight,College,Salary
5,Trent Forrest,Atlanta Hawks,G,6-abr,210.0,Florida State,508891.0
15,Miles Norris,Atlanta Hawks,NaN,NaN,NaN,NaN,559782.0
10,Seth Lundy,Atlanta Hawks,G,6-jun,220.0,Penn State,559782.0
7,Mouhamed Gueye,Atlanta Hawks,F,6-nov,210.0,Washington State,1119563.0
11,Garrison Mathews,Atlanta Hawks,G,6-jun,215.0,Lipscomb,2000000.0
...,...,...,...,...,...,...,...
582,Kyle Kuzma,Washington Wizards,F,6-sept,221.0,Utah,25568182.0
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
578,Taj Gibson,Washington Wizards,F,6-sept,232.0,Southern California,NaN
584,Kendrick Nunn,Washington Wizards,G,6-mar,190.0,Oakland,NaN


,Name,Team,Position,Height,Weight,College,Salary
17,Trae Young,Atlanta Hawks,G,6-ene,164.0,Oklahoma,40064220.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
8,De'Andre Hunter,Atlanta Hawks,F-G,6-ago,221.0,Virginia,20089286.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
14,Dejounte Murray,Atlanta Hawks,G,6-may,180.0,Washington,18214000.0
...,...,...,...,...,...,...,...
585,Eugene Omoruyi,Washington Wizards,F,6-jun,235.0,Oregon,559782.0
572,Jared Butler,Washington Wizards,G,6-mar,193.0,Baylor,559782.0
578,Taj Gibson,Washington Wizards,F,6-sept,232.0,Southern California,NaN
584,Kendrick Nunn,Washington Wizards,G,6-mar,190.0,Oakland,NaN


,Name,Team,Position,Height,Weight,College,Salary
590,Delon Wright,Washington Wizards,G,6-may,185.0,Utah,8195122.0
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN
588,Landry Shamet,Washington Wizards,G,6-abr,190.0,Wichita State,10250000.0
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
586,Jordan Poole,Washington Wizards,G,6-abr,194.0,Michigan,27955357.0
...,...,...,...,...,...,...,...
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
2,Kobe Bufkin,Atlanta Hawks,G,6-may,195.0,Michigan,4094244.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0


## Asignar rangos con el método `rank`
El método `rank` asigna un rango numérico a cada valor en una **Series**.
- Valores iguales comparten el mismo rango, dejando "huecos" en la numeración.
- Es útil para comparar posiciones relativas, por ejemplo, salarios dentro de un equipo.


Retorna una serie con los valroes con indices en la posicion de un ranking.  Si el rankin se repite, implica empate pero contabiliza posiiones.

In [199]:
nba = pd.read_csv(datasets_path+"/nba.csv", sep =";")
nba = nba.sort_values(["Team", "Name"])
nba

,Name,Team,Position,Height,Weight,College,Salary
6,AJ Griffin,Atlanta Hawks,F,6-jun,220.0,Duke,3712920.0
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0
8,De'Andre Hunter,Atlanta Hawks,F-G,6-ago,221.0,Virginia,20089286.0
...,...,...,...,...,...,...,...
587,Ryan Rollins,Washington Wizards,G,6-mar,180.0,Toledo,1719864.0
578,Taj Gibson,Washington Wizards,F,6-sept,232.0,Southern California,NaN
589,Tristan Vukcevic,Washington Wizards,F,6-oct,220.0,Real Madrid,NaN
580,Tyus Jones,Washington Wizards,G,6-feb,196.0,Duke,14000000.0


In [200]:
rank_serie = nba["Salary"].rank(ascending = True)
display(rank_serie)

#puedo ordenar acorde a este criterio

nba["ranking_salary"]=nba["Salary"].rank(ascending=False) #serie con indices originales y valroes posiciones. Uso las pocisiones como valores!
display(nba.sort_values(by = "ranking_salary", ascending = True))


6      233.0
1      409.0
4      181.0
3      420.0
8      417.0
       ...  
587     94.5
578      NaN
589      NaN
580    385.0
573     94.5
Name: Salary, Length: 591, dtype: float64

,Name,Team,Position,Height,Weight,College,Salary,ranking_salary
175,Stephen Curry,Golden State Warriors,G,6-feb,185.0,Davidson,51915615.0,1.0
461,Kevin Durant,Phoenix Suns,F,6-oct,240.0,Texas,47649433.0,2.0
145,Nikola Jokic,Denver Nuggets,C,6-nov,284.0,Mega Basket,47607350.0,4.0
436,Joel Embiid,Philadelphia 76ers,C-F,7-0,280.0,Kansas,47607350.0,4.0
261,LeBron James,Los Angeles Lakers,F,6-sept,250.0,St. Vincent-St. Mary HS (OH),47607350.0,4.0
...,...,...,...,...,...,...,...,...
543,Otto Porter Jr.,Toronto Raptors,F,6-ago,198.0,Georgetown,NaN,NaN
537,Ron Harper Jr.,Toronto Raptors,F,6-may,245.0,Rutgers,NaN,NaN
584,Kendrick Nunn,Washington Wizards,G,6-mar,190.0,Oakland,NaN,NaN
578,Taj Gibson,Washington Wizards,F,6-sept,232.0,Southern California,NaN,NaN


## Funcion apply

Aplica una funcion como condicion a una columna

In [201]:
#Ejemplo con condicionales de rango de salarios
def categorize_salary(salary):
    if salary > 20000000:
        return "Alto"
    elif salary > 5000000:
        return "Medio"
    else:
        return "Bajo"

nba["Salary Category"] = nba["Salary"].apply(categorize_salary)
nba.head()

,Name,Team,Position,Height,Weight,College,Salary,ranking_salary,Salary Category
6,AJ Griffin,Atlanta Hawks,F,6-jun,220.0,Duke,3712920.0,256.0,Bajo
1,Bogdan Bogdanovic,Atlanta Hawks,G,6-may,225.0,Fenerbahce,18700000.0,80.0,Medio
4,Bruno Fernando,Atlanta Hawks,F-C,6-oct,240.0,Maryland,2581522.0,308.0,Bajo
3,Clint Capela,Atlanta Hawks,C,6-oct,256.0,Elan Chalon,20616000.0,69.0,Alto
8,De'Andre Hunter,Atlanta Hawks,F-G,6-ago,221.0,Virginia,20089286.0,72.0,Alto
